ecdsa/src/rfc6979.rs

```diff
@@ -28,82 +29,47 @@ where
     Scalar<C>:
         PrimeField<Repr = FieldBytes<C>> + FromDigest<C> + Invert<Output = Scalar<C>> + Zeroize,
 {
-    let mut x = secret_scalar.to_repr();
-    let h1 = Scalar::<C>::from_digest(msg_digest).to_repr();
-    let mut hmac_drbg = HmacDrbg::<D>::new(&x, &h1, additional_data);
+    let bits_to_bytes = |b: elliptic_curve::ScalarBits<C>| {
+        // make byte buffer of the right size
+        let mut tmp = FieldBytes::<C>::default();
+        // convert bitslice from internal repr based on u32 or u64 to one based on u8
+        tmp.as_mut_slice()
+            .view_bits_mut::<Lsb0>()
+            .clone_from_bitslice(b.as_bitslice());
+        // convert to big endian
+        tmp.reverse();
+        tmp
+    };
+
+    // Verify that values roundtrip through bits_to_bytes and from_repr as expected.
+    // The from_repr interface does not guarantee big endian interpretation, and if it is not
+    // as expected the generated value may not have the required entropy.
+    let one = Scalar::<C>::one();
+    let one_bytes = bits_to_bytes(one.to_le_bits());
+    assert!(
+        one_bytes[one_bytes.len() - 1] == 1 && Scalar::<C>::from_repr(one_bytes) == Some(one),
+        "curve repr not as expected"
+    );
+
+    // Convert inputs to big endian byte representation
+    let modulus = bits_to_bytes(Scalar::<C>::char_le_bits());
+    let h1 = bits_to_bytes(Scalar::<C>::from_digest(msg_digest).to_le_bits());
+    let mut x = bits_to_bytes(secret_scalar.to_le_bits());
```

Comment:
All of this feels like complexity that's trying to work around the potential that `to_repr` will ever be anything other than big endian.

That will *never* be the case for ECDSA, which is to say that ECDSA is undefined for any curve with a non-big-endian field element representation, and the chances of that changing at any point in the future is, for all intents and purposes, zero.

If you'd prefer to have a separate trait defined by the `ecdsa` crate which by contract will always emit a big endian-serialized `FieldBytes`, we can define that, but really I don't think there's any need for any of this, and separately there shouldn't be any need to loop in any sort of bit-level views of field elements.

Reply:
I agree that ECDSA with the wrong kind of representation does not make sense. I just want to make sure it explodes instead of silently doing the wrong thing.

I also agree on not wanting to deal with `bitvec`. The endian thing is not the only reason why I need it though! The `char_le_bits` API is the only way to get the modulus of the prime field, and it returns a `bitvec` based type too.

Reply:
Also, worth pointing out that there is already an ECDSA specific trait: `FromDigest`. Perhaps it would be better to turn it into an `ECDSACompatible` trait or something, and add "must have a big endian repr" to its contract? I kind of hate the name, but just to get the idea across.

Reply:
> I agree that ECDSA with the wrong kind of representation does not make sense. I just want to make sure it explodes instead of silently doing the wrong thing.

The requirements for integrating the `ecdsa` crate with a particular curve implementation are a pretty tall order right now. We made the API deliberately coarse grained both to allow hiding certain things from the public API (i.e. base field elements) but also permitting optimizations that are only possible by leveraging potentially unique features of curve arithmetic backends.

- https://docs.rs/ecdsa/0.10.2/ecdsa/hazmat/trait.SignPrimitive.html
- https://docs.rs/ecdsa/0.10.2/ecdsa/hazmat/trait.VerifyPrimitive.html

Concrete implementations end up looking like this:

- https://github.com/RustCrypto/elliptic-curves/blob/master/k256/src/ecdsa/sign.rs#L211-L237
- https://github.com/RustCrypto/elliptic-curves/blob/master/k256/src/ecdsa/verify.rs#L110-L130

All of these things are explicitly called out as `hazmat` in the API and the bar is pretty high for even getting a particular implementation working in the first place, so I'm not too particularly worried that someone will manage to produce an ECDSA implementation which is otherwise working but manages to get everything else but endianness wrong.

Reply:
Ah, that makes a lot of sense. I had seen those traits briefly because I am just starting to think about what kind of public interface I want for the `dsa` code (it started from a very different place, with everything dynamic in the `rpgp` style instead of the mostly static dispatch style more common in RustCrypto), but had not really gone through them or considered why they are the way they are. This is good background information also for that work.

I'll take some time to digest this and then see how far I get making something we're both happy with.